src/utils/file.cpp

```diff
@@ -233,25 +271,108 @@ namespace file_util {
     return ret;
   }
 
+  /*
+   * Return the path without the last component (see dirname(3))
+   */
+  const string dirname(const string& path) {
```

Comment:
Why don't we use `dirname(3)` for this directly?

Reply:
I'd rather not reimplement a standard library function in our own codebase, even if using it is not so straightforward.

We could use `file_util::dirname(const string&)` as a wrapper around `dirname` that handles memory allocation, copying of the string, and thread safety.

In pseudocode, I think it would look something like this:
```
copy arg to newly allocated char*
lock
dirname(...)
copy result to std::string
unlock
free allocated char*

Reply:
I understand the desire to avoid reinventing the wheel, but the current implementation is only 11 lines of code, and the logic to use the real `dirname` is almost as long and, in my opinion, more complicated (in addition to being slower). Also, even with the lock, it won't be thread-safe if `dirname(3)` is used elsewhere in the executable (for example, by a statically linked library).

Here's a version that uses dirname directly. If you prefer it, I'll put it in.
``` cpp
const string dirname(string path) {
  static std::mutex m;

  // dirname may return a pointer to static memory,
  // use a lock to avoid race condition due to concurrent calls
  std::unique_lock<std::mutex> lock{m};
  auto ptr = dirname(&path[0]);

  if (ptr == path.c_str()) // modified in place
  {
    // dirname does not use static memory, release lock
    lock.unlock();

    path.erase(std::strlen(ptr));
    return std::move(path);
  }

  // cannot use erase here: for example, "usr" maps to "."
  path = ptr; // avoids an extra allocation
  return std::move(path);
}
```